journalpump/journalpump.py

```diff
@@ -727,28 +800,88 @@ def _close_stale_readers(self):
+
+        return False
+
+    def unregister_from_poll(self, reader: JournalReader) -> None:
+        fd = reader.fileno()
+        if fd is not None and fd in self.reader_by_fd:
+            self.poller.unregister(fd)
+            self.reader_by_fd[fd] = self._STALE_FD
+            self.log.info("Unregistered reader %r with fd %r", self.name, fd)
 
     def run(self):
         last_stats_time = 0
+        poll_timeout = 0
+        buffered_events = {}
+
         while self.running:
             self._close_stale_readers()
 
-            results = self.poller.poll(1000)
+            self.log.debug("Waiting for %dms", poll_timeout)
+            results = self.poller.poll(poll_timeout)
+            iteration_start_time = time.monotonic_ns()
+
             hits = {}
             lines = 0
-            for fd, _event in results:
-                for reader in self.readers.values():
-                    jdr = reader.journald_reader
-                    if not jdr or fd != jdr.fileno():
-                        continue
-                    if jdr.process() == systemd.journal.APPEND:
-                        lines += self.read_all_available_messages(reader, hits)
-                    break
-                else:
+            # We keep valid events in case reader is busy with processing
+            # previous batch and not ready to act on new one
+
+            for fd, _ in results:
+                reader = self.reader_by_fd.get(fd)
+                if reader is self._STALE_FD:
+                    continue
+
+                if reader is None:
                     self.log.error("Could not find reader with fd %r", fd)
+                    continue
+
+                # Call to process clears state, consecutive calls won't return same value
+                process_evt = reader.journald_reader.process()
+                # Can be either 0 - journal.NOP, 1 - journal.APPEND or 2 - journal.INVALIDATE
+                current_evt = buffered_events.get(reader, journal.NOP)
+                buffered_events[reader] = max(current_evt, process_evt)
+
+            # Remove stale reader, it should happen after we processed
+            # batch of events received from poll, because we could
+            # have received stale fds
+            for key, value in list(self.reader_by_fd.items()):
+                if value is self._STALE_FD:
+                    del self.reader_by_fd[key]
+
+            for reader, event in list(buffered_events.items()):
+                if not reader.is_ready:
+                    continue
+
+                # Read messages up to set reader limit
+                exhausted = False
+                if event in [journal.APPEND, journal.INVALIDATE]:
+                    exhausted, line_count = self.read_messages(reader, hits, chunk_size=5000)
+                    lines += line_count
+
+                if event == journal.INVALIDATE:
+                    self.unregister_from_poll(reader)
+                    reader.invalidate()
```

Comment:
If this isn't exhausted, doesn't this mean it doesn't read all? Should this be done only if exhausted?